File: Programmers/Algorithm_Kit/04_sort/003.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm> //sort
using namespace std;

int solution(vector<int> citations) {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);

    int answer = 0;

    sort(citations.begin(), citations.end(), greater<int>());

    if (citations[0] == 0) {
        answer = 0;
    }
    else {
        for (int i = 0; i < citations.size(); i++) {
            if (citations[i] >= i + 1) { //i + 1을 h로 가정 후 h-index 도출
                answer++;
            }
            else {
                break;
            }
        }
    }

    return answer;
}

File: programmers/algorithmkit/04_sorting/003.java
import java.util.*;

class Solution {
    public int solution(int[] citations) {
        int answer = 0;

        Arrays.sort(citations);

        // 배열 뒤집기
        for (int i = 0; i < citations.length / 2; i++) {
            int temp = citations[i];
            citations[i] = citations[citations.length - i - 1];
            citations[citations.length - i - 1] = temp;
        }

        if (citations[0] == 0) {
            answer = 0;
        } else {
            for (int i = 0; i < citations.length; i++) {
                if (citations[i] >= i + 1) {
                    answer++;
                } else {
                    break;
                }
            }
        }

        return answer;
    }
}

/**
 * 기본형은 Collections.reverseOrder() 사용 불가능
 * 
 * 역순으로 정렬하려면 래퍼 타입으로 변경하거나, 반복문으로 구현
 */